### ISSUE TITLE
spnego tokens with request flags fail to parse properly

### ISSUE THREAD

**Issue by wjackson** [jcmturner/gokrb5#390]
gokrb version: 8.3.0
go version: 1.14.3
relevant rfc: rfc2743 section 2.2.1 "req_flags"
client library: https://docs.spring.io/spring-security-kerberos/docs/1.0.1.RELEASE/reference/html/
bug demo code: https://pastebin.com/K170JCSR

When processing SPENGO tokens generated by the Java Spring Kerberos client library I see this error:

`continuation call to routine required`

It turns out that the problem tokens are not being properly parsed such that `SPNEGO.NegTokenInit.MechTokenBytes` is `nil`. That misleads spnego into thinking a continuation is needed.

The parser bug seems to be triggered by the fact that this client library sets request flags. Successful clients don't. When `ReqFlags` is present the asn1 struct parser fails to properly advance its byte array offset before attempting to parse the next field: `MechTokenBytes`. That results in the `MechTokenBytes` field being `nil`.

I suspect the issue is in the bit string parsing logic in `parseField` in `github.com/jcmturner/gofork@v1.0.0/encoding/asn1/asn1.go` . I don't know quite enough about how that's supposed to work to attempt a PR though.

The bug demo code linked to above shows the bug and also shows how to create bug triggering spnego tokens with request flags set.

**Comment by jcmturner** [comment]
Hi,
I seem unable to re-create the issue you are seeing. I have this test that tries to do the same thing as your pastebin code is doing but this seems to work OK. It's in the "investigation" branch: 
https://github.com/jcmturner/gokrb5/blob/investigation/v8/spnego/negotiationToken_test.go#L93-L114

**Comment by wjackson** [comment]
When using `NegTokenInit.Marshal` as you do in your unit test you end up with a serialized `NegTokenInit` that looks like this:

```
$ openssl asn1parse -in /tmp/yours -inform DER -i
    0:d=0  hl=2 l=  34 cons: cont [ 0 ]
    2:d=1  hl=2 l=  32 cons:  SEQUENCE
    4:d=2  hl=2 l=  13 cons:   cont [ 0 ]
    6:d=3  hl=2 l=  11 cons:    SEQUENCE
    8:d=4  hl=2 l=   9 prim:     OBJECT            :1.2.840.113554.1.2.2
   19:d=2  hl=2 l=   8 cons:   cont [ 1 ]
   21:d=3  hl=2 l=   6 cons:    SEQUENCE
   23:d=4  hl=2 l=   1 prim:     OCTET STRING      [HEX DUMP]:F6
   26:d=4  hl=2 l=   1 prim:     INTEGER           :07
   29:d=2  hl=2 l=   5 cons:   cont [ 2 ]
   31:d=3  hl=2 l=   3 prim:    OCTET STRING      [HEX DUMP]:010203
```

Unfortunately, the bug triggering `NetTokenInit`s in the wild look like this:

```
$ openssl asn1parse -in /tmp/mine -inform DER -i
    0:d=0  hl=2 l=  30 cons: cont [ 0 ]
    2:d=1  hl=2 l=  28 cons:  SEQUENCE
    4:d=2  hl=2 l=  13 cons:   cont [ 0 ]
    6:d=3  hl=2 l=  11 cons:    SEQUENCE
    8:d=4  hl=2 l=   9 prim:     OBJECT            :1.2.840.113554.1.2.2
   19:d=2  hl=2 l=   4 cons:   cont [ 1 ]
   21:d=3  hl=2 l=   2 prim:    BIT STRING
   25:d=2  hl=2 l=   5 cons:   cont [ 2 ]
   27:d=3  hl=2 l=   3 prim:    OCTET STRING      [HEX DUMP]:010203
```

To model the problem tokens, I needed to create and marshal `NetTokenInit` a bit more manually. I modified your unit test to do that here: https://github.com/wjackson/gokrb5/blob/f2d3590b31b94022789d271ed248239f0380ccac/v8/spnego/negotiationToken_test.go#L93-L153

Hopefully that helps.

**Comment by optiz0r** [comment]
I've been troubleshooting a similar problem with java client for a week or so before running across this ticket. It looks like I'm seeing the same behaviour. Wireshark inspection of the http packets show both flags and mechtoken present in a http request, but in negotiationToken.go:Verify `n.mechTokenBytes` is nil and authentication is failed. Our equivalent python code is working, which does not include any flags in the exchange.

The java code is using a lot of internal http and kerberos wrappers, so I can't easily post a simplified reproducer here.

**Comment by optiz0r** [comment]
Having spent most of the day on this with a colleague much more familiar with go than myself, we believe we have found the cause of this.

```go
type NegTokenInit struct {
	MechTypes      []asn1.ObjectIdentifier
	ReqFlags       gssapi.ContextFlags
	MechTokenBytes []byte
	MechListMIC    []byte
	mechToken      gssapi.ContextToken
	settings       *service.Settings
}

type marshalNegTokenInit struct {
	MechTypes      []asn1.ObjectIdentifier `asn1:"explicit,tag:0"`
	ReqFlags       gssapi.ContextFlags          `asn1:"explicit,optional,tag:1"`
	MechTokenBytes []byte                  `asn1:"explicit,optional,omitempty,tag:2"`
	MechListMIC    []byte                  `asn1:"explicit,optional,omitempty,tag:3"` // This field is not used when negotiating Kerberos tokens
}
```

Where `ReqFlags` has type `gssapi.ContextFlags` here, the call to `getUniversalType(fieldType)` for the reqFlags field will return universalTag=16 (sequence), compoundType=true instead of universalTag=3 (bitstring), compound=false. A but further on, there is a check that the tag about to be read from the byte stream has a class and tag matching the class and tag of the next field in the struct. `if t.class != expectedClass || t.tag != expectedTag || t.isCompound != compoundType {` This check fails because the tags 16!=3 and the compound true!=false. This causes the parseField to bail out without incrementing the offset, so the attempt to read the next NegToken struct field, MechTokenBytes is looking at the wrong offset within the buffer, and will also fail to be read. Since the MechTokenBytes field is left as nil, authentication subsequently fails.

We have found that changing the type of ReqFlags from `gssapi.ContextFlags` to `asn1.BitString` will cause `getUniversalType` to return the correct values, and so all fields for NegTokenInit are correctly parsed from the byte stream.

```go
type NegTokenInit struct {
	MechTypes      []asn1.ObjectIdentifier
	ReqFlags       asn1.BitString
	MechTokenBytes []byte
	MechListMIC    []byte
	mechToken      gssapi.ContextToken
	settings       *service.Settings
}

type marshalNegTokenInit struct {
	MechTypes      []asn1.ObjectIdentifier `asn1:"explicit,tag:0"`
	ReqFlags       asn1.BitString          `asn1:"explicit,optional,tag:1"`
	MechTokenBytes []byte                  `asn1:"explicit,optional,omitempty,tag:2"`
	MechListMIC    []byte                  `asn1:"explicit,optional,omitempty,tag:3"` // This field is not used when negotiating Kerberos tokens
}
```

As it stands, this workaround changes the semantics of the `ReqFlags` field, so is probably not acceptable as is. It just so happens that in our use case, we don't need to read from the ReqFlags. (I note from RFC4178 that this field was deprecated in 2005 and clients should omit it from the asn1 stream too).

The correct fix will be to ensure that the type `gssapi.ContextFlags` is interpreted by `getUniversalType` as a bitstream. It looks like it is already defined as such in `gssapi/contextFlags.go`, however this doesn't propagate through. In getUniversalType, the switch case fails to match this as bitStringType, and instead matches as a generic `reflect.Struct`.
```go
type ContextFlags asn1.BitString
```

**Comment by optiz0r** [comment]
I found a simpler way to handle this via type casting in PR #401. This compiles, tests pass, and tickets submitted by my Java client appear to work fine (there's nothing in the application on the server side that actually reads from the ContextFlags, so I'm only concerned with tickets being parsed correctly).

(I see asn1's `parseBitString` is already handling the big shifting, so no need for the changes suggested above. I also see that RFC4178 suggests the ContextFlags could be uint32 rather than 7bits padded in a single byte, so the bit-shifting code above is wrong anyway)

**Comment by jcmturner** [comment]
@optiz0r Thanks very much for the detailed investigation and the PR. I think the fix looks appropriate. What I'd like to do is have a test for this also. Do you think you'd be able to provide the bytes of a sample token that we could use to write a marshal/unmarshal test?

**Comment by optiz0r** [comment]
I spent enough time staring at pcaps that week, I'm pretty sure I can recall the bytes from memory!

The reqflags string from java 1.8 would be `0xa1 03 02 01 76`, and this appears right before the 0xa2 byte marking the start of the mechTokenBytes. Those bytes are indicating:
- tag1
- bitstring
- length 2
- padding bits 1
- flags 76 (7 flags in the top order bits, one trailing zero bit)

I can confirm from a pcap tomorrow or Monday.

**Comment by jcmturner** [comment]
Hi I've created PR #406 that also includes a test. I decided to just stop using the gssapi.ContextFlags type as this was only adding complexity for a field that is deprecated any way. 
Thanks for all your work on this in helping to find the root of this issue.

**Comment by jcmturner** [comment]
Sorry for the delay - this should now be fixed in v8.4.2
